### PR TITLE
add rt.tracegc

### PR DESCRIPTION
**Pull request by WalterBright**
Druntime part of https://github.com/D-Programming-Language/dmd/pull/4621

It's basically a stub, and will get fleshed out when the idea is accepted (i.e. pulled).

### CODE REVIEW

**Comment by WalterBright** [comment]
Fleshed it out.

File: src/rt/tracegc.d [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+ *    (See accompanying file LICENSE)
+ * Authors:   Walter Bright
+ * Source: $(DRUNTIMESRC src/rt/_tracegc.d)
+ */
+
+module rt.tracegc;
+
+import core.stdc.stdio;
+
+version (none)
```
Comment (andralex) [inline src/rt/tracegc.d]: put this in a unittest?

File: src/rt/tracegc.d [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+      }
+    }
+}
+
+extern (C) Object _d_newclass(const ClassInfo ci);
+extern (C) void[] _d_newarrayT(const TypeInfo ti, size_t length);
+extern (C) void[] _d_newarrayiT(const TypeInfo ti, size_t length);
+extern (C) void[] _d_newarraymTX(const TypeInfo ti, size_t[] dims);
+extern (C) void[] _d_newarraymiTX(const TypeInfo ti, size_t[] dims);
+extern (C) void* _d_newitemT(in TypeInfo _ti);
```
Comment (andralex) [inline src/rt/tracegc.d]: use `const` everywhere, `in` is just confusing

File: src/rt/tracegc.d [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+extern (C) void* _d_newitemiT(in TypeInfo _ti);
+
+extern (C) Object _d_newclassTrace(string file, int line, string funcname, const ClassInfo ci)
```
Comment (andralex) [inline src/rt/tracegc.d]: line is uint

**Comment by andralex** [comment]
I think issuing one printf per allocation will be quite onerous for all but the most trivial apps. If you want to make this a turnkey solution (no user-defined code needed), you need to collect things in a hashtable.

File: src/rt/tracegc.d (hunk 1 above)
Comment (WalterBright) [inline src/rt/tracegc.d]: won't really work as a unittest

**Comment by WalterBright** [comment]
I'll do the hashtable, this is just a start.

**Comment by andralex** [comment]
k

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
We already have a druntime internal malloc based hashtable.
https://github.com/D-Programming-Language/druntime/blob/1a7f82549aad11630776cce736b9bf8751e5e165/src/rt/util/container/hashtab.d

**Pull request merged**